### PR TITLE
Add Broker Parameters to Requests

### PR DESCRIPTION
**Pull request by pkanher617**
This PR is a follow up to #2428, and adds the parameters needed to broker requests for embedded applications to the requests in the `BrokerAuthorizationCodeClient` and `BrokerRefreshTokenClient`.

### CODE REVIEW

File: lib/msal-common/src/client/AuthorizationCodeClient.ts [hunk 1]
```diff
@@ -271,6 +219,58 @@ export class AuthorizationCodeClient extends BaseClient {
...
+        const correlationId = request.correlationId || this.config.cryptoInterface.createNewGuid();
+        parameterBuilder.addCorrelationId(correlationId);
+
+        if (!StringUtils.isEmpty(request.claims) || this.config.authOptions.clientCapabilities && this.config.authOptions.clientCapabilities.length > 0) {
+            parameterBuilder.addClaims(request.claims, this.config.authOptions.clientCapabilities);
+        }
+
+        return parameterBuilder.createQueryString();
+    }
+
```
Comment (pkanher617) [inline lib/msal-common/src/client/AuthorizationCodeClient.ts:273]: Moved with no change

File: lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * - scopes                 - Array of scopes the application is requesting access to.
+ * - claims                 - A stringified claims request which will be added to all /authorize and /token calls. When included on a silent request, cache lookup will be skipped and token will be refreshed.
+ * - authority              - Url of the authority which the application acquires tokens from.
+ * - correlationId          - Unique GUID set per request to trace a request end-to-end for telemetry purposes.
+ * - account                - Account entity to lookup the credentials.
+ * - forceRefresh           - Forces silent requests to make network calls if true.
+ * - resourceRequestMethod      - HTTP Request type used to request data from the resource (i.e. "GET", "POST", etc.).  Used for proof-of-possession flows.
+ * - resourceRequestUri         - URI that token will be used for. Used for proof-of-possession flows.
+ */
+export type BrokeredSilentFlowRequest = SilentFlowRequest & {
```
Comment (jasonnutter) [inline lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts:19]: Do these need to live in common? Or can they live in browser (until we know broker is needed for something like Node)?

File: lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { SilentFlowRequest } from "../SilentFlowRequest";
+
+/**
+ * SilentFlow parameters passed by the user to retrieve credentials silently
+ * - scopes                 - Array of scopes the application is requesting access to.
```
Comment (jasonnutter) [inline lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts:10]: We should look at these comments, I'm not a fan of the typedocs they generate, I think theres a better way. Plus, it is duplicitive to document `SilentFlowRequest` params here.

File: lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts:19]: They need to be in common because common needs to know about the embeddedAppClientId and brokerRedirectUri. There isn't really a way around that, and I'm not a fan of making them one-off function parameters, especially if things start getting added to the broker flow.

File: lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-common/src/request/broker/BrokeredSilentFlowRequest.ts:10]: Happy to discuss ways to improve the typedoc comments

File: lib/msal-angular/src/msal.service.ts [hunk 4]
```diff
@@ -56,7 +56,7 @@ export class MsalService extends UserAgentApplication {
                 this.setloginInProgress(false);
             }
             else if (this.getAcquireTokenInProgress()) {
-                broadcastService.broadcast("msal:acquireTokenFailure", msalError);
+                broadcastService.broadcast("msal:acquireToken", msalError);
```
Comment (tnorling) [inline lib/msal-angular/src/msal.service.ts:59]: Why was this changed?

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts [hunk 5]
```diff
@@ -3,34 +3,54 @@
...
+    /**
+     * 
+     */
+    async handleRedirectPromise(): Promise<BrokerAuthenticationResult | null> {
+        this.cachedBrokerResponse = super.handleRedirectPromise() as Promise<BrokerAuthenticationResult>;
+        const cachedResponse = (await this.cachedBrokerResponse);
+        // TODO: What to do in cases of multi-account in broker?
+        if (cachedResponse) {
+            this.brokerAccount = cachedResponse.account;
+            // return cachedResponse;
```
Comment (tnorling) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:48]: Should this be uncommented?

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts [hunk 6]
```diff
@@ -188,7 +210,11 @@ export class BrokerClientApplication extends ClientApplication {
      */
     private async brokeredPopupRequest(validMessage: BrokerAuthRequest, clientPort: MessagePort): Promise<void> {
         try {
-            const response: BrokerAuthenticationResult = (await this.acquireTokenPopup(validMessage.request as PopupRequest)) as BrokerAuthenticationResult;
+            const popupRequest = validMessage.request as BrokerPopupRequest;
+            popupRequest.redirectUri = validMessage.embeddedAppRedirectUri;
+            popupRequest.embeddedAppClientId = validMessage.embeddedClientId;
+            popupRequest.brokerRedirectUri = this.getRedirectUri();
+            const response: BrokerAuthenticationResult = (await this.acquireTokenPopup(popupRequest)) as BrokerAuthenticationResult;
```
Comment (tnorling) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:217]: Don't think you need this since you're casting the value

```suggestion
            const response = (await this.acquireTokenPopup(popupRequest)) as BrokerAuthenticationResult;
```

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts [hunk 7]
```diff
@@ -206,13 +232,27 @@ export class BrokerClientApplication extends ClientApplication {
...
+    private async brokeredSilentRequest(validMessage: BrokerAuthRequest, clientPort: MessagePort, account?: AccountInfo): Promise<void> {
         try {
-            const response: BrokerAuthenticationResult = (await this.acquireTokenByRefreshToken(validMessage.request as SilentFlowRequest)) as BrokerAuthenticationResult;
+            const silentRequest = validMessage.request as BrokerSilentRequest;
+            silentRequest.embeddedAppClientId = validMessage.embeddedClientId;
+            silentRequest.embeddedAppRedirectUri = validMessage.embeddedAppRedirectUri;
+            if (!silentRequest.account) {
+                silentRequest.account = account;
+            }
+            const response: BrokerAuthenticationResult = (await this.acquireTokenByRefreshToken(silentRequest)) as BrokerAuthenticationResult;
```
Comment (tnorling) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:243]:
```suggestion
            const response = (await this.acquireTokenByRefreshToken(silentRequest)) as BrokerAuthenticationResult;
```

File: lib/msal-browser/src/broker/msg/req/BrokerAuthRequest.ts [hunk 8]
```diff
@@ -3,35 +3,37 @@
...
         super(BrokerMessageType.AUTH_REQUEST);
         this.embeddedClientId = embeddedClientId;
+        this.embeddedAppRedirectUri = embeddedAppRedirectUri;
         this.interactionType = interactionType;
         this.request = request;
     }
 
     static validate(message: MessageEvent): BrokerAuthRequest| null {
         // First, validate message type
+        // eslint-disable-next-line no-console
```
Comment (tnorling) [inline lib/msal-browser/src/broker/msg/req/BrokerAuthRequest.ts:28]: Is this still needed?

File: lib/msal-browser/src/broker/msg/resp/BrokerRedirectResponse.ts [hunk 9]
```diff
@@ -3,9 +3,12 @@
  * Licensed under the MIT License.
  */
 
-import { BrokerMessage } from "./BrokerMessage";
-import { BrokerMessageType } from "../utils/BrowserConstants";
+import { BrokerMessage } from "../BrokerMessage";
+import { BrokerMessageType } from "../../../utils/BrowserConstants";
 
+/**
+ *
```
Comment (tnorling) [inline lib/msal-browser/src/broker/msg/resp/BrokerRedirectResponse.ts:10]: Empty comment block?

File: lib/msal-common/src/client/AuthorizationCodeClient.ts [hunk 10]
```diff
@@ -145,59 +145,7 @@ export class AuthorizationCodeClient extends BaseClient {
...
-
-        const correlationId = request.correlationId || this.config.cryptoInterface.createNewGuid();
-        parameterBuilder.addCorrelationId(correlationId);
-
-        if (!StringUtils.isEmpty(request.claims) || this.config.authOptions.clientCapabilities && this.config.authOptions.clientCapabilities.length > 0) {
-            parameterBuilder.addClaims(request.claims, this.config.authOptions.clientCapabilities);
-        }
-
-        return parameterBuilder.createQueryString();
+        return this.executePostToTokenEndpoint(authority.tokenEndpoint+"?dc=ESTS-PUB-WUS2-AZ1-TEST1", requestBody, headers, thumbprint);
```
Comment (tnorling) [inline lib/msal-common/src/client/AuthorizationCodeClient.ts:148]: Was the dc information intentionally committed? Need to make sure to remember to remove it before merging to dev...

File: lib/msal-common/src/client/RefreshTokenClient.ts [hunk 11]
```diff
@@ -135,7 +135,7 @@ export class RefreshTokenClient extends BaseClient {
             scopes: request.scopes
         };
 
-        return this.executePostToTokenEndpoint(authority.tokenEndpoint, requestBody, headers, thumbprint);
+        return this.executePostToTokenEndpoint(authority.tokenEndpoint+"?dc=ESTS-PUB-WUS2-AZ1-TEST1", requestBody, headers, thumbprint);
```
Comment (tnorling) [inline lib/msal-common/src/client/RefreshTokenClient.ts:138]: Same here

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 10 above)
Comment (tnorling) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:148]: I wonder if there's a better way we can provide test slices in the future that doesn't require changes to this code, as this isn't the first time we've had to do this and wont be the last

File: samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredWidget/script.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    });
+
+
+setTimeout(() => {
+    const contentElement = document.getElementsByClassName("myContent")[0];
+    contentElement.innerHTML = "I am now trying to fetch profile data";
+
+    setTimeout(async () => {
+        let exit = false;
+        await myMSALObj.ssoSilent({ loginHint: "dathayil@microsoft.com" })
```
Comment (tnorling) [inline samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredWidget/script.js:77]: Probably should remove hardcoded email?

File: lib/msal-angular/src/msal.service.ts (hunk 4 above)
Comment (pkanher617) [reply lib/msal-angular/src/msal.service.ts:59]: Not sure. I have reverted this change, may have occurred due to an errant merge resolution.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts (hunk 5 above)
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:48]: We only want to return here if the `cachedResponse` is for the broker. I will update this in a following PR.

Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:48]: I have updated the comment.

File: lib/msal-browser/src/broker/msg/resp/BrokerRedirectResponse.ts (hunk 9 above)
Comment (pkanher617) [reply lib/msal-browser/src/broker/msg/resp/BrokerRedirectResponse.ts:10]: fixed, updated comments for other APIs

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 10 above)
Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:148]: We currently don't allow extraQueryParameters on the token endpoint, which is why this fix was added. I agree we should add a feature to hit test slices, I will look into making a PR for this.

File: samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredWidget/script.js (hunk 12 above)
Comment (pkanher617) [reply samples/msal-browser-samples/BrokerTestApp2.0/app/brokeredWidget/script.js:77]: Removed in a later PR, but will also remove it here.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 10 above)
Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:148]: will remove before merging `broker` branch to dev.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts (hunk 6 above)
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts]: done.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts (hunk 7 above)
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts]: done.

File: lib/msal-browser/src/broker/msg/req/BrokerAuthRequest.ts (hunk 8 above)
Comment (pkanher617) [reply lib/msal-browser/src/broker/msg/req/BrokerAuthRequest.ts]: removed.

**Review by tnorling — APPROVED** [review]

**Review by jasonnutter — APPROVED** [review]

**Review by jmckennon — APPROVED** [review]

**Review by jo-arroyo — APPROVED** [review]
